File: app/src/main/java/com/bayanasar/inkverse/DiagReceiver.java
package com.bayanasar.inkverse;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

/**
 * adb-reachable control surface.
 *
 * The overlay covers the screen, so when a mode renders badly the on-screen buttons
 * are exactly what you cannot see. Everything needed to diagnose is therefore also
 * driveable from the host:
 *
 *   adb shell am broadcast -a com.bayanasar.inkverse.DIAG --es cmd mode --ei shader 0
 *   adb shell am broadcast -a com.bayanasar.inkverse.DIAG --es cmd probe
 *   adb shell am broadcast -a com.bayanasar.inkverse.DIAG --es cmd stop
 */
public class DiagReceiver extends BroadcastReceiver {

    public static final String ACTION = "com.bayanasar.inkverse.DIAG";

    @Override
    public void onReceive(Context ctx, Intent intent) {
        String cmd = intent.getStringExtra("cmd");
        Log.i(Tags.TAG, "DIAG cmd=" + cmd);
        if (cmd == null) return;

        switch (cmd) {
            case "start": {
                InvertAccessibilityService a = InvertAccessibilityService.get();
                if (a == null) { Log.w(Tags.TAG, "accessibility not running"); break; }
                String pkg = intent.getStringExtra("pkg");
                if (pkg != null) a.setTarget(pkg);
                a.start();
                break;
            }
            case "mode": {
                int shader = intent.getIntExtra("shader", 2);
                InvertAccessibilityService a = InvertAccessibilityService.get();
                if (a != null) a.setMode(shader);
                break;
            }
            case "target": {
                InvertAccessibilityService a = InvertAccessibilityService.get();
                if (a != null) a.setTarget(intent.getStringExtra("pkg"));
                break;
            }
            case "shape": {
                InvertAccessibilityService a2 = InvertAccessibilityService.get();
                if (a2 != null) {
                    a2.setShaping(intent.getFloatExtra("gamma", 1f),
                                  intent.getFloatExtra("black", 0f),
                                  intent.getFloatExtra("white", 1f));
                    break;
                }
                break;
            }

            case "probe": {
                InvertAccessibilityService a3 = InvertAccessibilityService.get();
                if (a3 != null) a3.probe();
                break;
            }
            case "stop": {
                InvertAccessibilityService a4 = InvertAccessibilityService.get();
                if (a4 != null) a4.stop();
                break;
            }
            default:
                Log.w(Tags.TAG, "unknown diag cmd: " + cmd);
        }
    }
}

File: app/src/main/kotlin/com/bayanasar/inkverse/DiagReceiver.kt
package com.bayanasar.inkverse

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log

/**
 * adb-reachable control surface.
 *
 * The overlay is opaque, so when a mode renders badly the on-screen controls are
 * exactly what you cannot see. Everything needed to diagnose is driveable from the
 * host instead:
 *
 *   adb shell am broadcast -a com.bayanasar.inkverse.DIAG -f 0x01000020 \
 *       --es cmd start --es pkg '*'
 *   adb shell am broadcast -a com.bayanasar.inkverse.DIAG -f 0x01000020 \
 *       --es cmd mode --ei shader 2
 *   adb shell am broadcast -a com.bayanasar.inkverse.DIAG -f 0x01000020 --es cmd probe
 *   adb shell am broadcast -a com.bayanasar.inkverse.DIAG -f 0x01000020 --es cmd stop
 *
 * The 0x01000020 flags are FLAG_INCLUDE_STOPPED_PACKAGES, needed because BOOX
 * auto-freeze leaves sideloaded apps in the stopped state.
 */
class DiagReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        val cmd = intent.getStringExtra("cmd") ?: return
        Log.i(TAG, "DIAG cmd=$cmd")

        val service = InvertAccessibilityService.get()
        if (service == null) {
            Log.w(TAG, "accessibility service is not running")
            return
        }

        when (cmd) {
            "start" -> {
                intent.getStringExtra("pkg")?.let(service::setTarget)
                service.start()
            }
            "stop" -> service.stop()
            "mode" -> service.setMode(intent.getIntExtra("shader", OverlayView.MODE_LUMA_INVERT))
            "target" -> service.setTarget(intent.getStringExtra("pkg"))
            "levels" -> service.setLevels(
                intent.getFloatExtra("black", 0f),
                intent.getFloatExtra("white", 1f),
            )
            "probe" -> service.probe()
            else -> Log.w(TAG, "unknown diag cmd: $cmd")
        }
    }
}
